service/src/main/java/org/mskcc/cmo/metadb/service/impl/AdminMessageHandlingServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.mskcc.cmo.messaging.Gateway;
+import org.mskcc.cmo.messaging.MessageConsumer;
+import org.mskcc.cmo.metadb.model.MetadbPatient;
+import org.mskcc.cmo.metadb.model.MetadbSample;
+import org.mskcc.cmo.metadb.model.SampleMetadata;
+import org.mskcc.cmo.metadb.service.AdminMessageHandlingService;
+import org.mskcc.cmo.metadb.service.CrdbMappingService;
+import org.mskcc.cmo.metadb.service.MetadbPatientService;
+import org.mskcc.cmo.metadb.service.MetadbSampleService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+/**
+ *
+ * @author ochoaa
+ */
+@Component
+public class AdminMessageHandlingServiceImpl implements AdminMessageHandlingService {
+
+    @Value("${mdb_admin.correct_cmoptid_topic}")
+    private String ADMIN_CORRECT_CMOPTID_TOPIC;
+
+    @Value("${mdb_admin.cmo_sample_label_update_topic}")
+    private String CMO_SAMPLE_LABEL_UPDATE;
+
+    @Value("${num.new_request_handler_threads}")
+    private int NUM_NEW_REQUEST_HANDLERS;
+
+    @Autowired
+    private CrdbMappingService crdbMappingService;
+
+    @Autowired
+    private MetadbPatientService patientService;
+
+    @Autowired
+    private MetadbSampleService sampleService;
+
+    private final ObjectMapper mapper = new ObjectMapper();
+    private static boolean initialized = false;
+    private static volatile boolean shutdownInitiated;
+    private static final ExecutorService exec = Executors.newCachedThreadPool();
+    private static final BlockingQueue<Map<String, String>> correctCmoPatientIdQueue =
+            new LinkedBlockingQueue<Map<String, String>>();
+    private static CountDownLatch correctCmoPatientIdShutdownLatch;
+    private static Gateway messagingGateway;
+
+    private static final Log LOG = LogFactory.getLog(AdminMessageHandlingServiceImpl.class);
+
+    private class CorrectCmoPatientIdReqReplyHandler implements Runnable {
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        CorrectCmoPatientIdReqReplyHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    Map<String, String> idCorrectionMap =
+                            correctCmoPatientIdQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (idCorrectionMap != null) {
+                        String oldCmoPatientId = idCorrectionMap.get("oldId");
+                        String newCmoPatientId = idCorrectionMap.get("newId");
+
+                        // get samples by old cmo patient id before updating the
+                        // cmo patient id for the given patient alias/patient node
+                        List<MetadbSample> samples =
```

Comment:
changed the return type here because `MetadbSample` contains `sampleCategory` that we can look at to determine if a sample is a research or clinical sample & handle them accordingly